### PR TITLE
Mirror shouldRunRecordFunction state in central TLS

### PR DESCRIPTION
**Pull request by robieta**
This PR will probably warrant some discussion. The current RecordFunction TLS is non-trivial (it contains a vector of classes), and is therefore not appropriate to move into the central TLS. That said, `shouldRunRecordFunction` doesn't need visibility into all of that detail; it just checks if there are any callbacks registered. This is a strawman which mirrors some invariants into the central TLS, with the idea that updates are infrequent and reuse of the TLS is useful. That said, this PR does violate Edward's law: "thou shalt not make an invalid state representable."

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59394 Mirror shouldRunRecordFunction state in central TLS**
* #59393 optimize checks
* #59392 move requires grad into central TLS
* #59391 move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h
* #59389 Add no-op benchmarks to test overhead

### CODE REVIEW

File: aten/src/ATen/core/dispatch/Dispatcher.h [hunk 1]
```diff
@@ -508,8 +508,9 @@ inline Return Dispatcher::callWithDispatchKeySlowPath(const TypedOperatorHandle<
 template<class Return, class... Args>
```
Comment (swolchok) [inline aten/src/ATen/core/dispatch/Dispatcher.h:508]: I will be surprised if making argument lists in these functions wider is worth whatever savings we're getting, particularly when we're statically linked and thus don't need tls_get_addr

**Review by swolchok — COMMENTED** [review]
personally, I'm not sold on this one. Duplicating state is a great way to create bugs, in addition to the likely cost of making the dispatcher have more live variables in its hot path. Would need to see what the wins are both in open-source and internally, preferably using perf profiling instead of instruction counting.

**Comment by robieta** [comment]
> personally, I'm not sold on this one. Duplicating state is a great way to create bugs, in addition to the likely cost of making the dispatcher have more live variables in its hot path. Would need to see what the wins are both in open-source and internally, preferably using perf profiling instead of instruction counting.

This would result in fewer hot variables, no? (Since in the common case we wouldn't actually look at the dedicated record function TLS and the generic TLS is already hot.) I should note that this implementation of sync at each call site is indeed bad; I just chose it because it was easy. A more robust impl would hide away the vector, and guard access with a consistency update. But I figured I should take the temperature about how palatable the concept was before diving too deep.

**Comment by ilia-cher** [comment]
Is the main reason we'd want this to avoid tls_get_addr? and moving the fields to the central place would only introduce these calls on the RecordFunction side?
Agree that we'd need to see some good perf result first to justify the duplication

**Comment by ilia-cher** [comment]
technically RecordFunction's tls relies on ThreadLocalState to preserve and reset it across the threads, if you'd prefer to keep these POD fields in the central TLS, then I think we could do this - we could make ThreadLocalState preserve/reset the whole POD local state, and also, if needed, set the duplicate tls fields here in record_function.cpp, then at least the duplicate fields are going to be handled in one place (in ThreadLocalState's ctor and setThreadLocalState)

**Comment by robieta** [comment]
This is indeed to skip the tls_get_addr. (And technically the vector::empty call, though this is a secondary effect.) The overall saving is 18 instructions which is not huge but also not trivial. (And there isn't going to be a single giant overhead win. We just have to shave bit by bit.) Basically I'm using the fact that shouldRunRecordFunction doesn't really need to know anything about the callbacks themselves. But I agree that the coupling needs to be more robust than this proof of concept if we decide to go forward with it.

**Pull request closed without merging**